File: runinator-command-center/src/core/utils/secrets.ts
import type { CredentialSummary } from "../domain/models";

export const SECRET_REF_PREFIX = "secret://";

export function secretKey(secret: CredentialSummary): string {
  return `${secret.kind ?? "secret"}:${secret.scope}:${secret.name}`;
}

export function secretRef(scope: string, name: string): string {
  return `${SECRET_REF_PREFIX}${encodeURIComponent(scope)}/${encodeURIComponent(name)}`;
}

// wdl-style reference for a setting, e.g. `secret.github.token` or `config.api.url`.
export function settingRef(
  kind: "secret" | "config" | undefined,
  scope: string,
  name: string,
): string {
  return `${kind ?? "secret"}.${scope}.${name}`;
}

export function parseSecretRef(value: unknown): CredentialSummary | null {
  if (typeof value !== "string" || !value.startsWith(SECRET_REF_PREFIX)) {
    return null;
  }

  const path = value.slice(SECRET_REF_PREFIX.length);
  const [rawScope, rawName] = path.split("/", 2);

  if (!rawScope || !rawName) {
    return null;
  }

  return {
    scope: decodeURIComponent(rawScope),
    name: decodeURIComponent(rawName),
  };
}

export function secretRefLabel(value: unknown): string {
  const parsed = parseSecretRef(value);
  return parsed ? `${parsed.scope}/${parsed.name}` : "";
}

File: runinator-command-center-flutter/lib/core/utils/secrets.dart
// port of core/utils/secrets.ts.

import '../domain/models/index.dart';

const String secretRefPrefix = 'secret://';

String secretKey(CredentialSummary secret) => '${secret.kind?.wire ?? 'secret'}:${secret.scope}:${secret.name}';

String secretRef(String scope, String name) =>
    '$secretRefPrefix${Uri.encodeComponent(scope)}/${Uri.encodeComponent(name)}';

/// wdl-style reference for a setting, e.g. `secret.github.token` or `config.api.url`.
String settingRef(SettingKind? kind, String scope, String name) => '${kind?.wire ?? 'secret'}.$scope.$name';

CredentialSummary? parseSecretRef(Object? value) {
  if (value is! String || !value.startsWith(secretRefPrefix)) {
    return null;
  }

  final path = value.substring(secretRefPrefix.length);
  final parts = path.split('/');

  if (parts.length < 2 || parts[0].isEmpty || parts[1].isEmpty) {
    return null;
  }

  return CredentialSummary(scope: Uri.decodeComponent(parts[0]), name: Uri.decodeComponent(parts[1]));
}

String secretRefLabel(Object? value) {
  final parsed = parseSecretRef(value);
  return parsed != null ? '${parsed.scope}/${parsed.name}' : '';
}
